### ISSUE TITLE
Auto-fill saves personal data on disk without the user's knowledge

### ISSUE THREAD

**Issue by fmarier** [brave/brave-browser#31960]
By default, the toggle in `brave://settings/addresses` is enabled:
![Screenshot from 2023-07-28 12-48-16](https://github.com/brave/brave-browser/assets/167821/f4a67b8b-4be9-4062-8a14-60f525a7dc85)

https://fmarier.org/test/autofill.html shows two different ways that this feature works:

1. *Incomplete address form*: not all of the fields are present and so the individual components will be saved in the browser (hidden, can't see or delete them individually) **without a prompt**.
2. *Complete address form*: all of the fields are present and a prompt will be shown before the complete address is saved and visible in `brave://settings/addresses`. If *denied*, the full address is not saved, but **the individual components are still saved** just as in the previous case (invisibly and without the user's knowledge).

The only indication that portions of an address are saved is in `brave://settings/clearBrowserData` (see _Auto-fill form data_) where all of the data for all sites can be deleted at once.

Note that nothing new is recorded in private windows, but the existing saved auto-fill info is available to be auto-filled in private browsing windows too (brave/brave-browser#9795).

**Comment by fmarier** [comment]
Saving **complete addresses** by default sounds good to me since there's a prompt each time (just like before passwords are saved), but the automatic and mostly-invisible saving of incomplete addresses by default seems problematic for a privacy browser.

**Comment by fmarier** [comment]
According to https://chromium.googlesource.com/chromium/src/+/main/components/autofill/README.md#what_s-the-difference-between-autofill-and-autocomplete, it looks like there is already a split in the code between _autofill_ (what I call _complete addresses_ above) and _autocomplete_ (what I refer to as _incomplete addresses_).

**Comment by fmarier** [comment]
Here's a concrete suggestion:

1. Add a new (default OFF) setting in `brave://settings/addresses` and alter the description of the existing setting:
![Screenshot from 2023-08-18 15-24-34](https://github.com/brave/brave-browser/assets/167821/b23dbbee-3599-4155-978d-652afe1be23c)

2. Add a new doorhanger prompt (like the "Save address?" one pictured above) when we're about to save an autocomplete field:

> Would you like Brave to save this kind of information and fill it in for you later?
> person@example.com
>
> [ Yes ] [ No ]

Clicking _Yes_ means we save the field (i.e. existing behavior) **and** flip the new setting to ON. Clicking _No_ means we don't save the field and we **never display this doorhanger again**.

**Comment by szilardszaloki** [comment]
Hey @fmarier,
Not sure how wild of a thought it is, but don't we want to disable autocomplete entirely on the autofill path? IIUC, the user is prompted each time they're about to save any autofill/typed information (addresses, credit/debit cards, etc.). If the user hits `No, thanks`, it doesn't seem to make a whole lot of sense to store the info as autocomplete data (not even if the user agreed to do so via the other prompt we're just about to introduce). If the user says they don't want the data to be saved, they most probably mean "not in any way", rather than "not as autofill data, but autocomplete is okay".

**Comment by fmarier** [comment]
> IIUC, the user is prompted each time Chromium is about to save any autofill/typed information (addresses, credit/debit cards, etc.).

I was actually thinking (though I didn't make that clear, thanks for pointing it out) that it would be a one-time prompt. You have to go into the settings to enable/disable later.

So amending my earlier suggestion slightly:

> Would you like Brave to save this kind of information and fill it in for you later?
> person@example.com
> 
> [ Yes, always ] [ No, never ]

That said, I'm open to other people's opinions. I was mainly trying to put something concrete out there to move this forward and encourage comments.

> If the user hits `No, thanks`, it doesn't seem to make a whole lot of sense to store the info as autocomplete data (not even if the user agreed to do so via the other prompt we're just about to introduce). If the user says they don't want the data to be saved, they most probably mean "not in any way", rather than "not as autofill data, but autocomplete is okay". WDYT?

I think you make a good point. If you say _no_ to saving parts of an address, then you most likely don't want to save that complete address either.

If we make the auto-fill prompt one-time only, then perhaps the double-prompt is not a big deal. On the other hand, one could say that if we make the auto-fill prompt one-time only, then it's also not a big deal if you can't save the very first complete address (but you get prompted to save any future ones).

Maybe @aguscruiz has some thoughts here.

**Comment by szilardszaloki** [comment]
@fmarier I know you were thinking that ours would be a one-time prompt. 🙂 I should have worded better. So what I meant is that if the user says "no" (via Chromium's built-in autofill prompt) to saving the complete address (as autofill data), then they most likely don't want to save parts of the address (as autocomplete data) - that's what I meant by "disabling autocomplete entirely on the autofill path". And therefore it wouldn't really matter what the user's choice was on our prompt (if they've seen our prompt at all), as we would get rid of the autocomplete functionality when the autofill path is triggered (that is, when saving typed data, e.g. addresses, credit/debit cards, IBAN addresses, etc.). This would also mean that the double-prompt could never occur.

**Comment by fmarier** [comment]
So if I understand your suggestion correctly, this is what the two cases would look like (assuming user says yes to prompts):

- incomplete address form:
  1. individual fields get saved into the auto-complete DB
  2. nothing gets saved into the auto-fill "address" DB
- complete address form:
  1. no individual fields are saved into the auto-complete DB
  2. the entire address is saved in the auto-fill "address" DB

So you're right that it would eliminate the possibility of a double prompt. However, it would also mean that users who said yes to the autocomplete prompt would also miss out on some data that would otherwise be remembered when filling out full address forms.

**Comment by szilardszaloki** [comment]
Yeah, and that's why I was wondering if this would be problematic above. It does seem to be redundant, as the data is already saved as autofill data, but I'm not exactly sure about the UX they would miss in that case.

**Comment by szilardszaloki** [comment]
But anyway, it would probably be best to not change this behavior.

**Comment by szilardszaloki** [comment]
I did check the double-prompt by the way, and it's not the end of the world.

**Comment by fmarier** [comment]
@szilardszaloki @aguscruiz and myself just had a chat and decided that the following approach would make the most sense:

- Add a new setting (default **off**) for _autocomplete_ at the root level in `brave://settings/autofill` and rename "Addresses and more" to just "Addresses":
![Screenshot from 2023-10-19 11-12-30](https://github.com/brave/brave-browser/assets/167821/90f429e7-c6e8-4606-9ac6-84827be5ee02)
- Update the description of the existing toggle in `brave://settings/addresses`:
![Screenshot from 2023-10-19 11-14-16](https://github.com/brave/brave-browser/assets/167821/d22c47ec-2a0b-4939-a08a-adc8894738e4)
- Show a one-off [prompt](https://github.com/brave/design/issues/345) before saving the first autocomplete field.
- Prevent autocomplete from saving anything until the user has opted in and the setting is **on**.
- If a Chromium autofill prompt is shown (e.g. when a complete address form is submitted), then suppress our new prompt to avoid confusing users with two popups.

**Comment by will-gant** [comment]
Would love to see some changes here. I think my use case is pretty common, but it's not currently well catered to:

1. I use an external password manager that has a browser extension, and want this (and only this) to autofill usernames and passwords
1. I want my address autofilled by the browser.

At the moment it appears the only way to achieve this is by toggling 'Save and fill addresses', _but_ the consequence of that is that whenver I focus on any login prompt I get:
a) A suggested login from my password manager (always the thing what I want)
b) A separate autofill suggestion from Brave (which, rather irritatingly, tends to float above the password manager's suggestion), proposing three or four usernames - many of which aren't usernames at all.

The only alternative to the above seems to be to disable 'Save and fill addresses', which means I lose the ability to have Brave fill my home address in forms.

**Comment by Brave-Matt** [comment]
Chiming in here -- we get a good amount of users asking if there is a way to completely disable autofill functionality. As it stands, even with any/all "autofill" based settings disabled, the browser will still retain form data and subsequently suggest it to the user when they enter a form field. They can delete this data (manually or via the `Delete browsing data` menu), but that is just a band-aid fix, as the next time they enter form data, that will be retained as well. Further, some users may want to autofill addresses, but not autofill other types of form data and currently both are tied to the same toggle option.

I think it might be worth expanding this issue to allow users to disable any/all autofill behavior, if desired. ideally it would be on a per type basis (maybe toggles for each type of data -- form data, payment data, passwords, etc.) but I think a "disable all autofill" option would be better than nothing.

Related issue:
https://github.com/brave/brave-browser/issues/32048

Some examples from Community:
https://community.brave.com/t/how-to-make-brave-never-remember-anything-i-type-into-a-field/604359/2
https://community.brave.com/t/is-this-browsing-history-or-autofill-form-data-and-how-do-i-clear-it/517844/5
https://community.brave.com/t/auto-suggest-should-not-automatically-fill-the-first-option-by-default/546454
https://community.brave.com/t/disable-search-suggestions-autofill/510437
https://community.brave.com/t/auto-fill-turned-off-but-walmart-website-autofills/493064/8
https://community.brave.com/t/unwanted-com-autofill-in-address-line/477195/4
https://community.brave.com/t/settings-autofill-password-manager-never-saved-sites-still-saving-user-name-how-to-stop-it/458459/9
https://community.brave.com/t/how-can-i-disable-the-drop-down-menu-that-suggests-auto-fill-data-for-usernames-and-passwords/534140/5

**Comment by robbycbennett** [comment]
Problem 1: we cannot disable autofill without disabling payments, addresses, etc.

Problem 2: autofill shared across different websites. Currently, autofill values are saved and loaded by the HTML `name` attribute. This means that if websites have the same name such as `<input name="search">` or `<input name="q">` those suggestions will be saved and shared for all websites. This is not ideal.

I would prefer to disable disabled entirely and enable it for a few exceptions. For instance, I would want to use autofill for my library card number on my county library website and prevent it from showing up in other websites.
